### ISSUE TITLE
System.exit when starting and stopping and unneccessary threads in Service class

### ISSUE THREAD

**Issue by fdmoulin** [perwendel/spark#731]
A few issues that are related, first ones are most urgent...

1 - Initializing a [spark.Service](https://github.com/perwendel/spark/blob/master/src/main/java/spark/Service.java) instance can result in a System.exit() call when the embedded Jetty server fails to start (E.g. on a port bind exception).  Service.stop() also calls System.exit() when an exception is thrown by the Jetty server stop method.  This means that Spark API users can have their applications "hard stopped" when using these methods, making them dangerous.

Offending snippet in [EmbeddedJettyServer.java](https://github.com/perwendel/spark/blob/master/src/main/java/spark/embeddedserver/jetty/EmbeddedJettyServer.java) :
```java
        try {
            logger.info("== {} has ignited ...", NAME);
            logger.info(">> Listening on {}:{}", host, port);

            server.start();
            latch.countDown();
            server.join();   //Not needed
        } catch (Exception e) {
            logger.error("ignite failed", e);
            System.exit(100);   //Not a good idea!!!
        }

return port;
```

Ideally, if a spark.Service instance's underlying Jetty server fails to start, the Spark API user should have a chance to catch an exception and react accordingly in their application.  The solution is to add a throws clause to the EmbeddedJettyServer's ignite() method and pass its exception down the chain to the spark.Service invoking method and NOT do System.exit.  It would be great if the Service.init method would rethrow this exception, but that would mean an API change that could break things for existing users.  I'd suggest to at least get the exception down to the Service class and log it there.  But again, the correct way to handle this would be to deprecate the existing Service.init method and replace it with one that throws the exception back to the API user.

2 - I've noticed something else that seems odd from an implementation standpoint.  The spark.Service.init() method creates a new thread that anonymously invokes the EmbeddedJettyServer's ignite() method, which calls join() on the Jetty server thread (shown above).  This is not necessary; the invoking thread could start the server if the EmbeddedJettyServer's ignite() method omitted the join().  In fact, because the join() is called, I don't believe the ignite method will ever return with its "int port" until the Jetty server shuts down, or in the event that it fails to start.  The Service.stop() method also uses the same technique of creating a new thread to clear routes and invoke the EmbeddedJettyServer's extinguish() method.  This new thread isn't necessary either since the underlying Jetty server's stop method isn't expected to block indefinitely.  In fact it's better to wait until it returns, ensuring that the server is properly stopped before freeing the invoking thread to take further actions that may put the system in an unstable state or cause errors.

Example from Service.stop method (same thing is done in the Service.init method)
```java
public synchronized void stop() {
        new Thread(() -> {     //Why are we creating a new Thread?
            if (server != null) {
                routes.clear();
                server.extinguish();
                latch = new CountDownLatch(1);
            }

            staticFilesConfiguration.clear();
            initialized = false;
        }).start();
}
```

3 - I noticed a small weirdness in the EmbeddedJettyServer.ignite() method.  The logging messages are out of order.  They log that the server has started prior to actually invoking the "start method".  This is minor but could cause confusion when debugging.

Snippet:
```java
            logger.info("== {} has ignited ...", NAME);
            logger.info(">> Listening on {}:{}", host, port);

            server.start();
```

4 - Other issues that should probably be addressed at some point: the spark.Service.init() method is effectively an "init+start" method rolled into one.  Ideally this would be separated.  Another nitpick; the init+start method is always called when adding a route or a filter.  As an API user, I'd prefer to have the ability to do initialization separately from execution.  And because the init method is synchronized, this causes all addRoute/addFilter calls to wait on the same lock (not a huge issue though).  I get why it was done this way, to streamline the API language for the static Spark use cases, but anyway, just a thought.  I realize that changing these would cause a change in the public facing API.. so I don't recommend it until the next major version some day maybe? :)

In the meantime, I strongly recommend addressing points 1 and 2 (and 3 while we're at it).  Any thoughts?

**Comment by jakaarl** [comment]
Agree, there's a lot of asynchronous stuff where it isn't really needed (or even desired). Problem is, changing the behaviour of `Service.init()` and/or `Service.stop()` would be a rather intrusive change, potentially breaking things for a lot of users.

And yes, the port isn't returned, just like you said - see #714 .

**Comment by fdmoulin** [comment]
Just saw this: https://github.com/perwendel/spark/issues/563 which explains why a new Thread was added to the stop method.

**Comment by asanchezktz** [comment]
I understend why is in a thread, but should be a way to know when stops. Just like `Spark.awaitInitialization()`. 

I'm using spock for the functional tests and i made this workaround to make it work:
```groovy
    private waitUntilServerStops() {
        while (true) {
            try {
                Spark.port()
                Thread.sleep(100)
            } catch (IllegalStateException ignored) {
                break
            }
        }
    }
```
